rfcs/80-curve-trait.md

```diff
@@ -0,0 +1,793 @@
+easings, animations, geometry, camera moves, and more!
+
+The trait itself looks like this:
+
+```rust
+pub trait Curve<T>
+{
+    fn domain(&self) -> Interval;
+    fn sample(&self, t: f32) -> T;
+}
+```
+At a basic level, it encompasses values of type `T` parametrized over some particular `Interval` (its `domain`,
+allowing that data to be sampled by providing the associated time parameter. 
+
+### Intervals
+
+The `Interval` type plays a part of the definition of each `Curve`. This is a type which
+represents a nonempty closed interval over the `f32` numbers whose endpoints may be at infinity. Its provided methods are
+mostly self-explanatory:
+```rust
+/// Create a new [`Interval`] with the specified `start` and `end`. The interval can be infinite
+/// but cannot be empty; invalid parameters will result in an error.
+pub fn new(start: f32, end: f32) -> Result<Self, InvalidIntervalError> { //... }
+
+/// Get the start of this interval.
+pub fn start(self) -> f32 { //... }
+
+/// Get the end of this interval.
+pub fn end(self) -> f32 { //... }
+
+/// Get the length of this interval. Note that the result may be infinite (`f32::INFINITY`).
+pub fn length(self) -> f32 { //... }
+
+/// Returns `true` if this interval is finite.
+pub fn is_finite(self) -> bool { //... }
+
+/// Returns `true` if `item` is contained in this interval.
+pub fn contains(self, item: f32) -> bool { //... }
+
+/// Create an [`Interval`] by intersecting this interval with another. Returns an error if the
+/// intersection would be empty (hence an invalid interval).
+pub fn intersect(self, other: Interval) -> Result<Interval, InvalidIntervalError> { //... }
+
+/// Clamp the given `value` to lie within this interval.
+pub fn clamp(self, value: f32) -> f32 { //... }
+
+/// Get the linear map which maps this curve onto the `other` one. Returns an error if either
+/// interval is infinite.
+pub fn linear_map_to(self, other: Self) -> Result<impl Fn(f32) -> f32, InfiniteIntervalError> { //... }
+
+/// Get an iterator over equally-spaced points from this interval in increasing order.
+/// Returns an error if `points` is less than 2 or if the interval is unbounded.
+pub fn spaced_points(
+    self,
+    points: usize,
+) -> Result<impl Iterator<Item = f32>, SpacedPointsError> {
+```
+
+The `Interval` type also implements `TryFrom<RangeInclusive>`, which may be desirable if you want to use
+the `start..=end` syntax. One of the primary benefits of `Interval` (in addition to these methods) is
+that it is `Copy`, so it is easy to take intervals and throw them around. 
+
+### Sampling
+
+The `Curve::sample` method is not intrinsically constrained by the curve's `domain` interval. Instead,
+implementors of `Curve<T>` are free to determine how samples drawn from outside the `domain` will behave.
+However, variants of `sample` (as well as other important methods) use the `domain` explicitly:
+```rust
```

Comment:
Suggestion: I think this section / API will be clearer if we explicitly provide an `extrapolate` method.